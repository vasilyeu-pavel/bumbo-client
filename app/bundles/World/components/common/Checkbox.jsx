import React, { Component, PropTypes } from 'react';
import { bindAll, map } from 'lodash';
import classNames from 'classnames';

export default class Checkbox extends Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    checked: PropTypes.bool,
    labelText: PropTypes.string,
    onChange: PropTypes.func,
    modifiers: PropTypes.array,
  }

  static defaultProps = {
    checked: false,
    modifiers: [],
  }

  constructor(props, context) {
    super(props, context);

    this.state = { checked: this.props.checked };

    bindAll(this, 'handleChange');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== this.props.checked) {
      this.setState({ checked: nextProps.checked });
    }
  }

  handleChange(event) {
    this.props.onChange(this.props.id, !this.state.checked);
    this.setState({ checked: !this.state.checked });
  }

  getClassNames() {
    return classNames('checkbox', map(this.props.modifiers,
      (modifier) => `checkbox--${modifier}`));
  }

  render() {
    return (
      <div className={this.getClassNames()}>
        <input
          ref={'checkbox'}
          className='checkbox__real'
          id={this.props.id}
          type='checkbox'
          checked={this.state.checked}
          onChange={this.handleChange}
        />
        <label className='checkbox__label' htmlFor={this.props.id}>
          <div className='checkbox__icon'></div>
          <div className='checkbox__label-text'>
            {this.props.labelText}
            <div className='checkbox__label-caption'>{this.props.captionText}</div>
          </div>
        </label>
      </div>
    );
  }
}
