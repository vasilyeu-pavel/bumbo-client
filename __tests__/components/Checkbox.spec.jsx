/* global expect, jest */

import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Checkbox from '../../app/bundles/World/components/common/Checkbox';

describe('<Checkbox /> component', () => {
  const minProps = {
    id: 'filter',
    labelText: 'filterEnabled',
  };

  it('should render input checkbox with label', () => {
    const checkbox = renderer.create(<Checkbox {...minProps} />);
    const checkboxTree = checkbox.toJSON();

    expect(checkboxTree).toMatchSnapshot();
  });

  it('should render input checkbox and "form-field" class', () => {
    const checkbox = renderer.create(<Checkbox {...minProps} modifiers={['form-field']} />);
    const checkboxTree = checkbox.toJSON();

    expect(checkboxTree).toMatchSnapshot();
  });

  it('should call the "onChange" method and set checked to true after input change', () => {
    const handleChange = (id, isChecked) => {
      expect(id).toEqual(minProps.id);
      expect(isChecked).toEqual(true);
    };
    const handler = jest.fn(handleChange);
    const checkbox = renderer.create(<Checkbox {...minProps} onChange={handler} />);
    let checkboxTree = checkbox.toJSON();

    const input = checkboxTree.children.filter(child => child.type === 'input')[0];

    input.props.onChange();

    expect(handler).toHaveBeenCalled();

    checkboxTree = checkbox.toJSON();

    expect(checkboxTree).toMatchSnapshot();
  });

  it('should change the "checked" prop after receiving new one', () => {
    const checkbox = mount(<Checkbox {...minProps}/>);

    checkbox.setProps({ checked: false });

    expect(checkbox.prop('checked')).toEqual(false);

    checkbox.setProps({ checked: true });

    expect(checkbox.prop('checked')).toEqual(true);
  });
});
