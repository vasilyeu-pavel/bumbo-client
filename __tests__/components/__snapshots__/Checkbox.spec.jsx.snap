exports[`<Checkbox /> component should call the "onChange" method and set checked to true after input change 1`] = `
<div
  className="checkbox ">
  <input
    checked={true}
    className="checkbox__real"
    id="filter"
    onChange={[Function]}
    type="checkbox" />
  <label
    className="checkbox__label"
    htmlFor="filter">
    <div
      className="checkbox__icon" />
    <div
      className="checkbox__label-text">
      filterEnabled
    </div>
  </label>
</div>
`;

exports[`<Checkbox /> component should render input checkbox and "form-field" class 1`] = `
<div
  className="checkbox checkbox--form-field">
  <input
    checked={false}
    className="checkbox__real"
    id="filter"
    onChange={[Function]}
    type="checkbox" />
  <label
    className="checkbox__label"
    htmlFor="filter">
    <div
      className="checkbox__icon" />
    <div
      className="checkbox__label-text">
      filterEnabled
    </div>
  </label>
</div>
`;

exports[`<Checkbox /> component should render input checkbox with label 1`] = `
<div
  className="checkbox ">
  <input
    checked={false}
    className="checkbox__real"
    id="filter"
    onChange={[Function]}
    type="checkbox" />
  <label
    className="checkbox__label"
    htmlFor="filter">
    <div
      className="checkbox__icon" />
    <div
      className="checkbox__label-text">
      filterEnabled
    </div>
  </label>
</div>
`;
